lib/pbench/server/hierarchy.py

```diff
@@ -0,0 +1,406 @@
+""" Collection and Structure Module for pbench-audit-server
+"""
+
+import os
+import time
+
+from collections import OrderedDict
+
+
+permdict = {
+    "0": "---",
+    "1": "--x",
+    "2": "-w-",
+    "3": "-wx",
+    "4": "r--",
+    "5": "r-x",
+    "6": "rw-",
+    "7": "rwx",
+}
+
+
+def filepermissions(mode):
+    """Convert File permissions from numeric to symbolic.
+    from '755' --> 'rwxr-xr-x' """
+    fperm = ""
+    modebits = oct(mode)[-3:]
+    for bits in modebits:
+        try:
+            fperm += permdict[bits]
+        except IndexError:
+            raise ValueError(
+                f"Could not find key '{bits}' in file permissions dictionary"
+            )
+
+    return fperm
+
+
+class Hierarchy(object):
+    """Super class of Hierarchies"""
+
+    def __init__(self, name, path, config):
+        self.name = name
+        self.path = path
+        self.config = config
+
+    def header(self, fp, status):
+        """Adds main starting and ending messages for each Hierarchy"""
+        lead_newline = "\n" if status == "start" else ""
+        fp.write(
+            f"{lead_newline}\n{status}-{self.config.TS[4:]}: {self.name} hierarchy: {self.path}\n"
+        )
+        return
+
+    def check_controller_in_list(self, controller, hierarchy_check):
```

Comment:
Note that "hierarchy_check" is actually a class variable in each of the subclasses. Although you used different names in each subclass, that's not really necessary; and if the property was promoted to a base class property (e.g., self.validation_list), you wouldn't need to pass the argument in multiple places. That would make cleaner and more cohesive code.

Reply:
yes it did.. thanks